go-controller/pkg/ovn/egressip.go

```diff
@@ -361,6 +361,120 @@ type egressIPMode struct {
 	gatewayIPCache sync.Map
 }
 
+func (e *egressIPMode) addPodEgressIP(eIP *egressipv1.EgressIP, pod *kapi.Pod) error {
+	podIPs := e.getPodIPs(pod)
+	if podIPs == nil {
+		e.podRetry.Store(getPodKey(pod), true)
+		return nil
+	}
+	if e.needsRetry(pod) {
+		e.podRetry.Delete(getPodKey(pod))
+	}
+	for _, status := range eIP.Status.Items {
+		if err := e.createEgressPolicy(podIPs, status, eIP.Name); err != nil {
+			return fmt.Errorf("unable to create logical router policy for status: %v, err: %v", status, err)
+		}
+		if err := createNATRule(podIPs, status, eIP.Name); err != nil {
+			return fmt.Errorf("unable to create NAT rule for status: %v, err: %v", status, err)
+		}
+	}
+	return nil
+}
+
+func (e *egressIPMode) deletePodEgressIP(eIP *egressipv1.EgressIP, pod *kapi.Pod) error {
+	podIPs := e.getPodIPs(pod)
+	if podIPs == nil {
+		return nil
+	}
+	for _, status := range eIP.Status.Items {
+		if err := e.deleteEgressPolicy(podIPs, status, eIP.Name); err != nil {
+			return fmt.Errorf("unable to delete logical router policy for status: %v, err: %v", status, err)
+		}
+		if err := deleteNATRule(podIPs, status, eIP.Name); err != nil {
+			return fmt.Errorf("unable to delete NAT rule for status: %v, err: %v", status, err)
+		}
+	}
+	return nil
+}
+
+func createNATRule(podIPs []net.IP, status egressipv1.EgressIPStatusItem, egressIPName string) error {
+	for _, podIP := range podIPs {
+		if (utilnet.IsIPv6String(status.EgressIP) && utilnet.IsIPv6(podIP)) || (!utilnet.IsIPv6String(status.EgressIP) && !utilnet.IsIPv6(podIP)) {
+			natIDs, err := findNatIDs(egressIPName, podIP.String(), status.EgressIP)
+			if err != nil {
+				return err
+			}
+			if natIDs == nil {
+				_, stderr, err := util.RunOVNNbctl(
+					"--id=@nat",
+					"create",
+					"nat",
+					"type=snat",
+					fmt.Sprintf("logical_port=k8s-%s", status.Node),
+					fmt.Sprintf("external_ip=%s", status.EgressIP),
+					fmt.Sprintf("logical_ip=%s", podIP),
+					fmt.Sprintf("external_ids:name=%s", egressIPName),
+					"--",
+					"add",
+					"logical_router",
+					fmt.Sprintf("GR_%s", status.Node),
+					"nat",
+					"@nat",
+				)
+				if err != nil {
+					return fmt.Errorf("unable to create nat rule, stderr: %s, err: %v", stderr, err)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func deleteNATRule(podIPs []net.IP, status egressipv1.EgressIPStatusItem, egressIPName string) error {
+	for _, podIP := range podIPs {
+		if (utilnet.IsIPv6String(status.EgressIP) && utilnet.IsIPv6(podIP)) || (!utilnet.IsIPv6String(status.EgressIP) && !utilnet.IsIPv6(podIP)) {
+			natIDs, err := findNatIDs(egressIPName, podIP.String(), status.EgressIP)
```

Comment:
instead of doing 2 operations here to find the nat ID and then delete the nat ID, you could just do one lr-nat-del since you already know what you are deleting. Can take care of this in a follow up though.

Reply:
No that won't work, because: https://github.com/ovn-org/ovn-kubernetes/commit/3a0c54c2fd66034d8b3b851673bfb3a83129e76e